src/types/viewport.cpp

```diff
@@ -295,6 +300,61 @@ bool Viewport::WalkInBounds(til::point& pos, const til::CoordType delta, bool al
     return off == offClamped;
 }
 
+bool Viewport::IncrementInExclusiveBounds(til::point& pos) const noexcept
+{
+    return WalkInExclusiveBounds(pos, 1);
+}
+
+bool Viewport::DecrementInExclusiveBounds(til::point& pos) const noexcept
+{
+    return WalkInExclusiveBounds(pos, -1);
+}
+
+bool Viewport::IsInExclusiveBounds(const til::point pos) const noexcept
+{
+    return pos.x >= Left() && pos.x <= RightExclusive() &&
+           pos.y >= Top() && pos.y <= BottomInclusive();
+}
+
+int Viewport::CompareInExclusiveBounds(const til::point first, const til::point second) const noexcept
+{
+    // Assert that our coordinates are within the expected boundaries
+    assert(IsInExclusiveBounds(first));
+    assert(IsInExclusiveBounds(second));
+
+    // First set the distance vertically
+    //   If first is on row 4 and second is on row 6, first will be -2 rows behind second * an 80 character row would be -160.
+    //   For the same row, it'll be 0 rows * 80 character width = 0 difference.
+    auto retVal = (first.y - second.y) * Width();
+
+    // Now adjust for horizontal differences
+    //   If first is in position 15 and second is in position 30, first is -15 left in relation to 30.
+    retVal += (first.x - second.x);
+
+    // Further notes:
+    //   If we already moved behind one row, this will help correct for when first is right of second.
+    //     For example, with row 4, col 79 and row 5, col 0 as first and second respectively, the distance is -1.
+    //     Assume the row width is 80.
+    //     Step one will set the retVal as -80 as first is one row behind the second.
+    //     Step two will then see that first is 79 - 0 = +79 right of second and add 79
+    //     The total is -80 + 79 = -1.
+    return retVal;
+}
+
+bool Viewport::WalkInExclusiveBounds(til::point& pos, const til::CoordType delta) const noexcept
+{
+    const auto l = static_cast<ptrdiff_t>(_sr.left);
+    const auto t = static_cast<ptrdiff_t>(_sr.top);
+    const auto w = static_cast<ptrdiff_t>(std::max(0, _sr.right - _sr.left + 2));

```

Comment:
I think this should be +1. BTW is this function really needed? We already have `WalkInBounds` with the `allowEndExclusive` parameter that we could set to `false` no?

Reply:
> I think this should be +1. BTW is this function really needed? We already have `WalkInBounds` with the `allowEndExclusive` parameter that we could set to `false` no?

Reply:
Ah, so `allowEndExclusive` is _slightly_ different. `EndExclusive = (Left, BottomExclusive)`. `allowEndExclusive` was designed to make `WalkInBounds(true)` still walk from `Left` --> `RightInclusive`, but when we get to the last cell of the bottom inclusive row, we wrap onto the next line (EndExclusive). This made it so that we're able to represent the UIA start/end as an exclusive range, and we'd be able to include the last cell of the buffer by making end = EndExclusive.

In hindsight, I think the right move should've been to make `EndExclusive = (RightExclusive, BottomInclusive)`. This PR makes selection do that and adds these extra APIs. I think it's much cleaner. But yeah, I wouldn't want to port UIA over to this system until it gets reviewed and merged. So I'm making that work a part of #4423 as a follow-up (where we should be able to remove `EndExclusive` altogether).

As for the `+1`, I honestly don't remember why I made it `+2`... you're probably right. I'll make that change and validate.